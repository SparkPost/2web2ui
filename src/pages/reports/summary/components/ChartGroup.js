import React, { Component } from 'react';
import _ from 'lodash';
import { getDayLines, getLineChartFormatters } from 'src/helpers/chart';
import LineChart from './LineChart'; // async
import METRICS_UNIT_CONFIG from 'src/config/metrics-units';
const DEFAULT_UNIT = 'number';

function getUniqueUnits(metrics) {
  return _.uniq(metrics.map(({ unit = DEFAULT_UNIT }) => unit));
}

export default class ChartGroup extends Component {
  createDayReferenceLines() {
    const { chartData, precision } = this.props;

    return getDayLines(chartData, precision).map(({ ts }) => ({
      key: ts,
      x: ts,
      stroke: '#bbb',
      strokeWidth: 2,
    }));
  }

  render() {
    const { chartData = [], metrics, chartLoading, precision, yScale, to } = this.props;

    if (!chartData.length || !metrics) {
      return null;
    }

    const formatters = getLineChartFormatters(precision, to);
    const referenceLines = this.createDayReferenceLines();
    const charts = getUniqueUnits(metrics).map(unit => ({
      metrics: metrics.filter(metric => metric.unit === unit),
      ...METRICS_UNIT_CONFIG[unit],
    }));

    return (
      <div>
        {charts.map((chart, i) => (
          <LineChart
            key={`chart=${i}`}
            syncId="summaryChart"
            data={chartData}
            precision={precision}
            lines={chart.metrics.map(({ name, label, stroke }) => ({
              key: name,
              dataKey: name,
              name: label,
              stroke: chartLoading ? '#f8f8f8' : stroke,
            }))}
            {...formatters}
            yTickFormatter={chart.yAxisFormatter}
            yScale={yScale}
            yLabel={chart.label}
            tooltipValueFormatter={chart.yAxisFormatter}
            referenceLines={referenceLines}
            showXAxis={i === charts.length - 1}
          />
        ))}
      </div>
    );
  }
}
