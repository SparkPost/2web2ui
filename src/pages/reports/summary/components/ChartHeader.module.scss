@import '~@sparkpost/matchbox/src/styles/config.scss';

.ChartHeader {
  margin-bottom: spacing(large);
}

.TimeMode {
  display: inline-block;
  font-size: rem(13);
  line-height: rem(15);
  color: color(gray, 4);
  padding: spacing(small) 0 0;
}

.Controls {
  margin-top: spacing();
  margin-left: spacing(larger);
  text-align: left;

  & * {
    text-align: left;
  }

  @media screen and (min-width: breakpoint(small)) {
    margin-top: 0;
    margin-left: 0;
    text-align: right;
  }
}

.ButtonSpacer {
  margin-left: spacing(small);
}
