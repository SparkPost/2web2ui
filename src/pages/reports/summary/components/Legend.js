import React from 'react';
import { connect } from 'react-redux';
import getConfig from 'src/helpers/getConfig';
import styles from './Legend.module.scss';

function renderMetric(metric, uniqueLabel) {
  return (
    <div className={styles.Metric} key={metric.name}>
      {/* eslint-disable-next-line */}
      <span className={styles.Color} style={{ backgroundColor: metric.stroke }} />
      {metric.isUniquePerTimePeriod && uniqueLabel
        ? `${metric.label} ${uniqueLabel}`
        : metric.label}
    </div>
  );
}

export const Legend = props => {
  const { metrics, reportOptions } = props;
  const { precision } = reportOptions;

  const metricsRollupPrecisionMap = getConfig('metricsRollupPrecisionMap');
  const precisionObj = metricsRollupPrecisionMap.find(p => p.value === precision);
  return (
    <div className={styles.Legend}>
      {metrics &&
        metrics.map(metric => renderMetric(metric, precisionObj && precisionObj.uniqueLabel))}
    </div>
  );
};

const mapStateToProps = state => ({
  reportOptions: state.reportOptions,
});

export default connect(mapStateToProps)(Legend);
