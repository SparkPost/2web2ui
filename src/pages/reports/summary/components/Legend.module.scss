@import '~@sparkpost/matchbox/src/styles/config.scss';

.Legend {
  margin-top: rem(3);
  margin-left: spacing(larger);
}

.Metric {
  display: inline-block;
  vertical-align: middle;
  margin-right: spacing(large);
  font-weight: 400;
}

.Color {
  display: inline-block;
  vertical-align: middle;
  margin-right: rem(9);
  margin-top: -2px;
  height: 14px;
  width: 14px;
  border-radius: border-radius();
}
