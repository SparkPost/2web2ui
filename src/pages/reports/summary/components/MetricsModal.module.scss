@import '~@sparkpost/matchbox/src/styles/config.scss';

.Apply {
  float: left;
}

.Cancel {
  float: right;
}

.Metric {
  display: inline-block;
  margin-bottom: rem(3);
  min-width: rem(210);
  width: 33%;
  margin-right: -4px;
}
