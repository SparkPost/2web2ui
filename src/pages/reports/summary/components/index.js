export { default as ChartGroup } from './ChartGroup';
export { default as ChartHeader } from './ChartHeader';
export { default as Legend } from './Legend';
export { default as Table } from './Table';
export { default as MetricsModal } from './MetricsModal';
