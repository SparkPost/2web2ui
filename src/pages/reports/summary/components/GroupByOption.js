import React, { Component } from 'react';
import styles from './Table.module.scss';
import { Grid, Checkbox, Select } from 'src/components/matchbox';
import _ from 'lodash';
import { GROUP_CONFIG } from './tableConfig';

export class GroupByOption extends Component {
  state = {
    topDomainsOnly: true,
  };

  handleGroupChange = e => {
    this.props._getTableData({ groupBy: e.target.value });
  };

  handleDomainsCheckboxChange = () => {
    const topDomainsOnly = !this.state.topDomainsOnly;
    this.setState({ topDomainsOnly });
    const groupBy = topDomainsOnly ? 'watched-domain' : 'domain';
    this.props._getTableData({ groupBy });
  };

  getSelectOptions = () => {
    const { hasSubaccounts } = this.props;
    const { topDomainsOnly } = this.state;

    const filteredOptionsKeys = _.reduce(
      GROUP_CONFIG,
      (accumulator, value, key) => {
        if (
          (key === 'subaccount' && !hasSubaccounts) ||
          (key === 'domain' && topDomainsOnly) ||
          (key === 'watched-domain' && !topDomainsOnly)
        ) {
          return accumulator; // ignore
        }
        accumulator.push(key);
        return accumulator;
      },
      [],
    );

    const options = filteredOptionsKeys.map(key => ({
      value: key,
      label: GROUP_CONFIG[key].label,
    }));

    return options;
  };

  renderDomainsCheckbox() {
    const { tableLoading, groupBy } = this.props;
    const { topDomainsOnly } = this.state;

    //Only show 'Top Domains Only' checkbox when on the recipient domains grouping
    if (groupBy !== 'watched-domain' && groupBy !== 'domain') {
      return null;
    }
    return (
      <div className={styles.TopDomainsCheckbox}>
        <Checkbox
          id="watchedDomains"
          label="Top Domains Only"
          checked={topDomainsOnly}
          onChange={this.handleDomainsCheckboxChange}
          disabled={tableLoading}
        />
      </div>
    );
  }

  render() {
    const { groupBy, tableLoading } = this.props;
    return (
      <Grid>
        <Grid.Column xs={12} md={5} lg={4}>
          <Select
            label="Group By"
            options={this.getSelectOptions()}
            value={groupBy}
            disabled={tableLoading}
            onChange={this.handleGroupChange}
          />
        </Grid.Column>
        <Grid.Column xs={12} md={4} mdOffset={3} lg={3} lgOffset={5}>
          {this.renderDomainsCheckbox()}
        </Grid.Column>
      </Grid>
    );
  }
}

export default GroupByOption;
