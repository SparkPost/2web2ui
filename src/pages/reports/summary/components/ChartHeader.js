import React, { Component } from 'react';
import { Grid, Button } from 'src/components/matchbox';
import Legend from './Legend';

import styles from './ChartHeader.module.scss';

export default class ChartHeader extends Component {
  renderScaleButton(scale, label) {
    const { onScaleClick, selectedScale } = this.props;
    return (
      <Button size="small" primary={scale === selectedScale} onClick={() => onScaleClick(scale)}>
        {label}
      </Button>
    );
  }

  renderTimeMode() {
    const { selectedTime, onTimeClick } = this.props;
    return (
      <Button.Group className={styles.ButtonSpacer}>
        <Button primary={selectedTime === 'real'} onClick={() => onTimeClick('real')} size="small">
          Real Time
        </Button>
        <Button
          primary={selectedTime === 'injection'}
          onClick={() => onTimeClick('injection')}
          size="small"
        >
          Injection Time
        </Button>
      </Button.Group>
    );
  }

  render() {
    const { selectedMetrics, onMetricsToggle } = this.props;

    return (
      <Grid className={styles.ChartHeader}>
        <Grid.Column xs={12} md={9}>
          <Legend metrics={selectedMetrics} />
        </Grid.Column>
        <Grid.Column xs={12} md={3}>
          <div className={styles.Controls}>
            <Button size="small" onClick={onMetricsToggle}>
              Select Metrics
            </Button>
            {/* {this.renderTimeMode()} */}
            {/*<Button.Group className={styles.ButtonSpacer}>*/}
            {/*{this.renderScaleButton('linear', 'Linear')}*/}
            {/*{this.renderScaleButton('log', 'Log')}*/}
            {/*{this.renderScaleButton('pow', 'Pow')}*/}
            {/*</Button.Group>*/}
          </div>
        </Grid.Column>
      </Grid>
    );
  }
}
