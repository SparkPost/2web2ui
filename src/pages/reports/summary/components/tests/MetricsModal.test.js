import React from 'react';
import { shallow } from 'enzyme';
import MetricsModal from '../MetricsModal';
import { Checkbox } from 'src/components/matchbox';

// this uses the src/config/__mocks__/metrics.js file automatically
jest.mock('src/config/metrics');

describe('Component: Summary Chart Metrics Modal', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      open: true,
      onCancel: jest.fn(),
      onSubmit: jest.fn(),
      selectedMetrics: [],
    };
    wrapper = shallow(<MetricsModal {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should select metrics', () => {
    const checkboxes = wrapper.find(Checkbox);
    checkboxes.at(1).simulate('change');
    checkboxes.at(3).simulate('change');

    expect(wrapper).toMatchSnapshot();
  });

  it('should disable checkboxes when max selected', () => {
    const checkboxes = wrapper.find(Checkbox);
    wrapper.setProps({ maxMetrics: 2 });

    checkboxes.at(1).simulate('change');
    checkboxes.at(3).simulate('change');

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle submit', () => {
    const checkboxes = wrapper.find(Checkbox);
    const instance = wrapper.instance();

    checkboxes.at(1).simulate('change');
    checkboxes.at(3).simulate('change');

    instance.handleKeyDown({ key: 'Enter' });

    expect(props.onSubmit).toHaveBeenCalledWith(['count_injected', 'count_targeted']);
  });

  it('should do nothing on key events if not open ðŸ¤”', () => {
    const checkboxes = wrapper.find(Checkbox);
    const instance = wrapper.instance();

    checkboxes.at(1).simulate('change');
    checkboxes.at(3).simulate('change');

    wrapper.setProps({ open: false });
    instance.handleKeyDown({ key: 'Enter' });
    instance.handleKeyDown({ key: 'Escape' });

    expect(props.onSubmit).not.toHaveBeenCalled();
    expect(props.onCancel).not.toHaveBeenCalled();
  });
});
