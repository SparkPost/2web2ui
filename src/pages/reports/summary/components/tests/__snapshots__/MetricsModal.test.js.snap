// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Summary Chart Metrics Modal should disable checkboxes when max selected 1`] = `
<Modal.LEGACY
  onClose={[MockFunction]}
  open={true}
>
  <WindowEvent
    event="keydown"
    handler={[Function]}
  />
  <Panel.LEGACY>
    <Panel.LEGACY.Section>
      <h5>
        Select up to 5 metrics
      </h5>
      <div>
        <div
          className="Metric"
          key="count_accepted"
        >
          <Tooltip
            content="Messages an ISP or other remote domain accepted (less Out-of-Band Bounces)."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={true}
              id="count_accepted"
              label="Accepted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_injected"
        >
          <Tooltip
            content="Messages either injected to or received by SparkPost."
            dark={true}
          >
            <Checkbox
              checked={true}
              disabled={false}
              id="count_injected"
              label="Injected"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_sent"
        >
          <Tooltip
            content="Messages that SparkPost attempted to deliver, which includes both Deliveries and In-Band Bounces."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={true}
              id="count_sent"
              label="Sent"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_targeted"
        >
          <Tooltip
            content="Messages successfully injected into SparkPost as well as rejected by it."
            dark={true}
          >
            <Checkbox
              checked={true}
              disabled={false}
              id="count_targeted"
              label="Targeted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
      </div>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <Button
        className="Apply"
        disabled={false}
        onClick={[Function]}
        primary={true}
      >
        Apply Metrics
      </Button>
      <Button
        className="Cancel"
        onClick={[MockFunction]}
      >
        Cancel
      </Button>
    </Panel.LEGACY.Section>
  </Panel.LEGACY>
</Modal.LEGACY>
`;

exports[`Component: Summary Chart Metrics Modal should render 1`] = `
<Modal.LEGACY
  onClose={[MockFunction]}
  open={true}
>
  <WindowEvent
    event="keydown"
    handler={[Function]}
  />
  <Panel.LEGACY>
    <Panel.LEGACY.Section>
      <h5>
        Select up to 5 metrics
      </h5>
      <div>
        <div
          className="Metric"
          key="count_accepted"
        >
          <Tooltip
            content="Messages an ISP or other remote domain accepted (less Out-of-Band Bounces)."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_accepted"
              label="Accepted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_injected"
        >
          <Tooltip
            content="Messages either injected to or received by SparkPost."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_injected"
              label="Injected"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_sent"
        >
          <Tooltip
            content="Messages that SparkPost attempted to deliver, which includes both Deliveries and In-Band Bounces."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_sent"
              label="Sent"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_targeted"
        >
          <Tooltip
            content="Messages successfully injected into SparkPost as well as rejected by it."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_targeted"
              label="Targeted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
      </div>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <Button
        className="Apply"
        disabled={true}
        onClick={[Function]}
        primary={true}
      >
        Apply Metrics
      </Button>
      <Button
        className="Cancel"
        onClick={[MockFunction]}
      >
        Cancel
      </Button>
    </Panel.LEGACY.Section>
  </Panel.LEGACY>
</Modal.LEGACY>
`;

exports[`Component: Summary Chart Metrics Modal should select metrics 1`] = `
<Modal.LEGACY
  onClose={[MockFunction]}
  open={true}
>
  <WindowEvent
    event="keydown"
    handler={[Function]}
  />
  <Panel.LEGACY>
    <Panel.LEGACY.Section>
      <h5>
        Select up to 5 metrics
      </h5>
      <div>
        <div
          className="Metric"
          key="count_accepted"
        >
          <Tooltip
            content="Messages an ISP or other remote domain accepted (less Out-of-Band Bounces)."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_accepted"
              label="Accepted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_injected"
        >
          <Tooltip
            content="Messages either injected to or received by SparkPost."
            dark={true}
          >
            <Checkbox
              checked={true}
              disabled={false}
              id="count_injected"
              label="Injected"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_sent"
        >
          <Tooltip
            content="Messages that SparkPost attempted to deliver, which includes both Deliveries and In-Band Bounces."
            dark={true}
          >
            <Checkbox
              checked={false}
              disabled={false}
              id="count_sent"
              label="Sent"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
        <div
          className="Metric"
          key="count_targeted"
        >
          <Tooltip
            content="Messages successfully injected into SparkPost as well as rejected by it."
            dark={true}
          >
            <Checkbox
              checked={true}
              disabled={false}
              id="count_targeted"
              label="Targeted"
              onChange={[Function]}
            />
          </Tooltip>
        </div>
      </div>
    </Panel.LEGACY.Section>
    <Panel.LEGACY.Section>
      <Button
        className="Apply"
        disabled={false}
        onClick={[Function]}
        primary={true}
      >
        Apply Metrics
      </Button>
      <Button
        className="Cancel"
        onClick={[MockFunction]}
      >
        Cancel
      </Button>
    </Panel.LEGACY.Section>
  </Panel.LEGACY>
</Modal.LEGACY>
`;
